src/ast.cpp

```diff
@@ -508,68 +515,62 @@ namespace Sass {
-      (*copy)[0] = this->unify_with(rhs_0, ctx);
-      return copy;
+      // Compound_Selector_Ptr copy = SASS_MEMORY_NEW(Compound_Selector, *rhs);
+      rhs->at(0) = this->unify_with(rhs_0, ctx);
+      return rhs;
 
     }
     // else it's a tag name and a bunch of qualifiers -- just append them
-    SimpleSequence_Selector* cpy = SASS_MEMORY_NEW(ctx.mem, SimpleSequence_Selector, rhs->pstate());
-    if (name() != "*") (*cpy) << this;
-    (*cpy) += rhs;
-    return cpy;
+    // Compound_Selector_Ptr cpy = SASS_MEMORY_NEW(Compound_Selector, rhs->pstate());
+    if (name() != "*") rhs->elements().insert(rhs->begin(), this);
+    return rhs;
   }
 
-  SimpleSequence_Selector* Class_Selector::unify_with(SimpleSequence_Selector* rhs, Context& ctx)
+  Compound_Selector_Ptr Class_Selector_Ref::unify_with(Compound_Selector_Ptr rhs, Context& ctx)
   {
     rhs->has_line_break(has_line_break());
     return Simple_Selector::unify_with(rhs, ctx);
   }
 
-  SimpleSequence_Selector* Id_Selector::unify_with(SimpleSequence_Selector* rhs, Context& ctx)
+  Compound_Selector_Ptr Id_Selector_Ref::unify_with(Compound_Selector_Ptr rhs, Context& ctx)
   {
     for (size_t i = 0, L = rhs->length(); i < L; ++i)
     {
-      Simple_Selector* rhs_i = (*rhs)[i];
-      if (typeid(*rhs_i) == typeid(Id_Selector) && static_cast<Id_Selector*>(rhs_i)->name() != name()) {
-        return 0;
+      if (Id_Selector_Ptr sel = SASS_MEMORY_CAST(Id_Selector, rhs->at(i))) {
+        if (sel->name() != name()) return 0;
       }
     }
     rhs->has_line_break(has_line_break());
     return Simple_Selector::unify_with(rhs, ctx);
   }
 
-  SimpleSequence_Selector* Pseudo_Selector::unify_with(SimpleSequence_Selector* rhs, Context& ctx)
+  Compound_Selector_Ptr Pseudo_Selector_Ref::unify_with(Compound_Selector_Ptr rhs, Context& ctx)
   {
     if (is_pseudo_element())
     {
       for (size_t i = 0, L = rhs->length(); i < L; ++i)
       {
-        Simple_Selector* rhs_i = (*rhs)[i];
-        if (typeid(*rhs_i) == typeid(Pseudo_Selector) &&
-            static_cast<Pseudo_Selector*>(rhs_i)->is_pseudo_element() &&
-            static_cast<Pseudo_Selector*>(rhs_i)->name() != name())
-        { return 0; }
+        if (Pseudo_Selector_Ptr sel = SASS_MEMORY_CAST(Pseudo_Selector, rhs->at(i))) {
+          if (sel->is_pseudo_element() && sel->name() != name()) return 0;
+        }
       }
     }
     return Simple_Selector::unify_with(rhs, ctx);
   }
 
-  bool Attribute_Selector::operator< (const Attribute_Selector& rhs) const
+  bool Attribute_Selector_Ref::operator< (const Attribute_Selector& rhs) const
   {
+
     if (is_ns_eq(ns(), rhs.ns())) {
-      if (name() == rhs.name()) {
-        if (matcher() == rhs.matcher()) {
-          return value() < rhs.value();
-        } else { return matcher() < rhs.matcher(); }
-      } else { return name() < rhs.name(); }
+      return (name() < rhs.name()) || (matcher() < rhs.matcher())
```

Comment:
this is not equivalent the current behaviour i.e. when `name() > rhs.name()` and `matcher() < rhs.matcher()`